File: ios/Pulpe/Domain/Models/TransactionEnums.swift
import SwiftUI

/// Type of financial flow
enum TransactionKind: String, Codable, CaseIterable, Sendable {
    case income
    case expense
    case saving

    var label: String {
        switch self {
        case .income: AppLocale.string("Revenu")
        case .expense: AppLocale.string("Dépense")
        case .saving: AppLocale.string("Épargne")
        }
    }

    /// SF Symbol name for savings (bank icon, used everywhere)
    static let savingsIcon = "building.columns"

    var icon: String {
        switch self {
        case .income: "arrow.down.left"
        case .expense: "arrow.up.right"
        case .saving: Self.savingsIcon
        }
    }

    var color: Color {
        switch self {
        case .income: .financialIncome
        case .expense: .financialExpense
        case .saving: .financialSavings
        }
    }

    var descriptionPlaceholder: String {
        switch self {
        case .expense: AppLocale.string("Ex : Courses, Restaurant...")
        case .income: AppLocale.string("Ex : Salaire, Remboursement...")
        case .saving: AppLocale.string("Ex : Vacances, Fonds d'urgence...")
        }
    }

    /// The verb carries the tense: here a fact already on the account, on the
    /// forecast side an intention still to come. "Nouvelle dépense" named
    /// neither — it only said something was about to exist.
    ///
    /// Whole sentences rather than a verb interpolated with an article: French puts the
    /// verb first and German puts it last, so a "Noter %@" template would force the
    /// French word order on every language.
    var newTransactionTitle: String {
        switch self {
        case .expense: AppLocale.string("Noter une dépense")
        case .income: AppLocale.string("Noter un revenu")
        case .saving: AppLocale.string("Noter une épargne")
        }
    }

    var newBudgetLineTitle: String {
        switch self {
        case .expense: AppLocale.string("Prévoir une dépense")
        case .income: AppLocale.string("Prévoir un revenu")
        case .saving: AppLocale.string("Prévoir une épargne")
        }
    }

    var editBudgetLineTitle: String {
        switch self {
        case .expense: AppLocale.string("Modifier la dépense")
        case .income: AppLocale.string("Modifier le revenu")
        case .saving: AppLocale.string("Modifier l'épargne")
        }
    }

    /// For calculations, savings are treated as expenses
    var isOutflow: Bool {
        self == .expense || self == .saving
    }
}

/// Recurrence type for budget lines
enum TransactionRecurrence: String, Codable, CaseIterable, Sendable {
    case fixed
    case oneOff = "one_off"

    var label: String {
        switch self {
        case .fixed: AppLocale.string("Récurrent")
        // French says "Prévu" for this type AND for the planned aggregate; the other
        // languages split them (One-off vs Planned), so this sense needs its own key —
        // the bare "Prévu" key stays the aggregate (docs/I18N.md, deliberate divergence 1).
        case .oneOff: AppLocale.string("recurrence.oneOff")
        }
    }

    var longLabel: String {
        switch self {
        case .fixed: AppLocale.string("Tous les mois")
        case .oneOff: AppLocale.string("Une seule fois")
        }
    }

    var icon: String {
        switch self {
        case .fixed: "repeat"
        case .oneOff: "1.circle"
        }
    }
}

File: android/app/src/main/java/app/pulpe/android/domain/model/TransactionEnums.kt
package app.pulpe.android.domain.model

import androidx.compose.ui.graphics.Color
import app.pulpe.android.ui.theme.FinancialExpense
import app.pulpe.android.ui.theme.FinancialIncome
import app.pulpe.android.ui.theme.FinancialSavings
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Type of financial flow
 */
@Serializable
enum class TransactionKind {
    @SerialName("income")
    INCOME,
    @SerialName("expense")
    EXPENSE,
    @SerialName("saving")
    SAVING;

    val label: String
        get() = when (this) {
            INCOME -> "Revenu"
            EXPENSE -> "Dépense"
            SAVING -> "Épargne"
        }

    val icon: String
        get() = when (this) {
            INCOME -> "arrow_downward"
            EXPENSE -> "arrow_upward"
            SAVING -> "savings"
        }

    val color: Color
        get() = when (this) {
            INCOME -> FinancialIncome
            EXPENSE -> FinancialExpense
            SAVING -> FinancialSavings
        }

    val isOutflow: Boolean
        get() = this == EXPENSE || this == SAVING
}

/**
 * Recurrence type for budget lines
 */
@Serializable
enum class TransactionRecurrence {
    @SerialName("fixed")
    FIXED,
    @SerialName("one_off")
    ONE_OFF;

    val label: String
        get() = when (this) {
            FIXED -> "Récurrent"
            ONE_OFF -> "Prévu"
        }

    val longLabel: String
        get() = when (this) {
            FIXED -> "Tous les mois"
            ONE_OFF -> "Une seule fois"
        }

    val icon: String
        get() = when (this) {
            FIXED -> "repeat"
            ONE_OFF -> "looks_one"
        }
}
